### ISSUE TITLE
Project loader throws IndexOutOfRangeException for unmanaged libraries

### ISSUE THREAD

**Issue by natemcmaster** [aspnet/dnx#3201]
This error happens inconsistently on CoreCLR/Linux. Loading SQLite triggers this error from dnx.

```
System.IndexOutOfRangeException : Index was outside the bounds of the array.
      Stack Trace:
           at System.Collections.Generic.HashSet`1.AddIfNotPresent(T value)
           at Microsoft.Dnx.Runtime.Loader.ProjectAssemblyLoader.LoadUnmanagedLibrary(String name)
           at Microsoft.Dnx.Host.LoaderContainer.LoadUnmanagedLibrary(String name)
           at Microsoft.Dnx.Host.DefaultLoadContext.LoadUnmanagedLibrary(String name)
           at Microsoft.Dnx.Runtime.Loader.LoadContext.LoadUnmanagedDll(String unmanagedDllName)
           at System.Runtime.Loader.AssemblyLoadContext.ResolveUnmanagedDll(String unmanagedDllName, IntPtr gchManagedAssemblyLoadContext)
```
See https://travis-ci.org/aspnet/EntityFramework/jobs/93045776

**Comment by natemcmaster** [comment]
cc @moozzyk

**Comment by moozzyk** [comment]
Feels like a problem that might occurred if load happens from multiple threads and the hashset is in an inconsistent state

**Comment by natemcmaster** [comment]
So, an issue for corefx? Or should we be using a System.Collections.Concurrent here instead?

**Comment by moozzyk** [comment]
I think we need to use Concurrent in dnx. If the user is trying to use DllImport in multiple threads I don't think clr will/should try serialize it.

**Comment by natemcmaster** [comment]
This bug is still causing frequently CI failures for EF on Linux. https://travis-ci.org/aspnet/EntityFramework/builds/94071437

@moozzyk Is native library loading going to part dotnet cli?

**Comment by moozzyk** [comment]
@natemcmaster - I don't know where it is going to land. @davidfowl will know. I will try to fix this soon to unblock you.

**Comment by moozzyk** [comment]
Fixed in 181910604.
